File: solr/core/src/java/org/apache/solr/handler/admin/api/Documents.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.solr.handler.admin.api;

import jakarta.inject.Inject;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.apache.lucene.util.BytesRef;
import org.apache.lucene.util.BytesRefBuilder;
import org.apache.solr.api.JerseyResource;
import org.apache.solr.client.api.endpoint.DocumentApi;
import org.apache.solr.client.api.model.GetDocumentResponse;
import org.apache.solr.client.api.model.ListDocumentsResponse;
import org.apache.solr.common.SolrDocument;
import org.apache.solr.common.SolrException;
import org.apache.solr.common.SolrInputDocument;
import org.apache.solr.common.util.StrUtils;
import org.apache.solr.core.SolrCore;
import org.apache.solr.handler.component.RealTimeGetComponent;
import org.apache.solr.jersey.PermissionName;
import org.apache.solr.schema.FieldType;
import org.apache.solr.schema.SchemaField;
import org.apache.solr.security.PermissionNameProvider;

/**
 * V2 API for fetching documents as path-identified resources.
 *
 * <p>Unlike {@link RealTimeGetAPI} (the JAX-RS mirror of the v1 {@code /get} handler, which is
 * relied on internally by SolrCloud peer-sync machinery and so must keep its exact {@code
 * id}/{@code ids} query-param contract), this API calls {@link RealTimeGetComponent}'s underlying
 * document-lookup utilities directly rather than delegating to a request handler, and returns a
 * {@code 404} for a document that doesn't exist.
 *
 * @see DocumentApi
 */
public class Documents extends JerseyResource implements DocumentApi {

  private final SolrCore solrCore;

  @Inject
  public Documents(SolrCore solrCore) {
    this.solrCore = solrCore;
  }

  @Override
  @PermissionName(PermissionNameProvider.Name.READ_PERM)
  public GetDocumentResponse getDocument(String id) throws IOException {
    final var response = instantiateJerseyResponse(GetDocumentResponse.class);

    final SolrDocument doc = fetchDocument(id);
    if (doc == null) {
      throw new SolrException(SolrException.ErrorCode.NOT_FOUND, "No such document: " + id);
    }
    response.doc = SolrDocumentFieldConverter.toFieldMap(doc);
    return response;
  }

  @Override
  @PermissionName(PermissionNameProvider.Name.READ_PERM)
  public ListDocumentsResponse listDocuments(List<String> ids) throws IOException {
    final var response = instantiateJerseyResponse(ListDocumentsResponse.class);

    final List<Map<String, Object>> docs = new ArrayList<>();
    if (ids != null) {
      // Each raw "ids" value may itself be a comma-separated list (e.g. "ids=a,b,c"), matching
      // the same convention RealTimeGetComponent.IdsRequested uses for the v1-mirrored /get API.
      for (String rawIds : ids) {
        for (String id : StrUtils.splitSmart(rawIds, ",", true)) {
          final SolrDocument doc = fetchDocument(id);
          if (doc != null) {
            docs.add(SolrDocumentFieldConverter.toFieldMap(doc));
          }
        }
      }
    }

    response.docs = docs;
    response.numFound = docs.size();
    response.start = 0;
    response.numFoundExact = true;
    return response;
  }

  private SolrDocument fetchDocument(String id) throws IOException {
    final SchemaField idField = solrCore.getLatestSchema().getUniqueKeyField();
    final FieldType fieldType = idField.getType();
    final BytesRefBuilder idBytesBuilder = new BytesRefBuilder();
    fieldType.readableToIndexed(id, idBytesBuilder);
    final BytesRef idBytes = idBytesBuilder.get();

    final SolrInputDocument inputDoc =
        RealTimeGetComponent.getInputDocument(
            solrCore, idBytes, idBytes, null, null, RealTimeGetComponent.Resolution.DOC);
    if (inputDoc == null) {
      return null;
    }
    return RealTimeGetComponent.toSolrDoc(inputDoc, solrCore.getLatestSchema());
  }
}

File: solr/webapp/web/js/angular/controllers/documents.js
/*
 Licensed to the Apache Software Foundation (ASF) under one or more
 contributor license agreements.  See the NOTICE file distributed with
 this work for additional information regarding copyright ownership.
 The ASF licenses this file to You under the Apache License, Version 2.0
 (the "License"); you may not use this file except in compliance with
 the License.  You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
//helper for formatting JSON and others

var DOC_PLACEHOLDER = '<doc>\n' +
                '  <field name="id">change.me</field>\n' +
                '  <field name="title">change.me</field>\n' +
                '</doc>';

var JSON_DOC_PLACEHOLDER = '{\n' +
                '  "id": "change.me",\n' +
                '  "title": "change.me"\n' +
                '}';
var JSON_COMMAND_PLACEHOLDER = '{\n' +
                '  "add": {\n' +
                '    "doc": {\n' +
                '      "id": "change.me",\n' +
                '      "title": "change.me"\n' +
                '    }\n' +
                '  }\n' +
                '}';

solrAdminApp.controller('DocumentsController',
    function($scope, $routeParams, Luke, UpdateV2, FileUpload, Constants, ApiErrorHandler) {
        $scope.resetMenu("documents", Constants.IS_COLLECTION_PAGE);

        $scope.refresh = function () {
            Luke.schema({core: $routeParams.core}, function(data) {
                //TODO: handle dynamic fields
                delete data.schema.fields._version_;
                $scope.fields = Object.keys(data.schema.fields);
            });
            $scope.document = "";
            $scope.type = "json";
            $scope.commitWithin = 1000;
            $scope.overwrite = true;
        };

        $scope.refresh();

        $scope.changeDocumentType = function () {
            $scope.placeholder = "";
            if ($scope.type == 'json') {
                $scope.placeholder = JSON_DOC_PLACEHOLDER;
            } else if ($scope.type == 'csv') {
                $scope.placeholder = "id,title\nchange.me,change.me";
            } else if ($scope.type == 'solr-json') {
                $scope.placeholder = JSON_COMMAND_PLACEHOLDER;
            } else if ($scope.type == 'xml') {
                $scope.placeholder = DOC_PLACEHOLDER;
            }
        };

        $scope.addWizardField = function () {
            if ($scope.document == "") $scope.document = "{}";
            var doc = JSON.parse($scope.document);
            doc[$scope.fieldName] = $scope.fieldData;
            $scope.document = JSON.stringify(doc, null, '\t');
            $scope.fieldData = "";
        };

        $scope.submit = function () {
            if ($scope.type == "upload") {
                FileUpload.upload({
                    core: $routeParams.core,
                    handler: "update",
                    commitWithin: $scope.commitWithin,
                    overwrite: $scope.overwrite,
                    wt: "json",
                    raw: $scope.literalParams
                }, $scope.fileUpload, function (data) {
                    $scope.responseStatus = "success";
                    $scope.response = JSON.stringify(data, null, '  ');
                }, function (data) {
                    $scope.responseStatus = "failure";
                    $scope.response = JSON.stringify(data, null, '  ');
                });
                return;
            }

            var postData;
            var updateMethod;
            if ($scope.type == "json" || $scope.type == "wizard") {
                postData = "[" + $scope.document + "]";
                updateMethod = UpdateV2.update;
            } else if ($scope.type == "solr-json") {
                postData = $scope.document;
                updateMethod = UpdateV2.update;
            } else if ($scope.type == "xml") {
                postData = "<add>" + $scope.document + "</add>";
                updateMethod = UpdateV2.updateXml;
            } else if ($scope.type == "csv") {
                postData = $scope.document;
                updateMethod = UpdateV2.updateCsv;
            }
            if (!updateMethod || $scope.isCloudEnabled === undefined) return;

            var indexType = $scope.isCloudEnabled ? "collections" : "cores";
            var updateOptions = {
                commitWithin: $scope.commitWithin,
                overwrite: $scope.overwrite
            };
            var v2Callback = function (error, data, response) {
                if (error) {
                    $scope.responseStatus = "failure";
                    $scope.response = JSON.stringify((response && response.body) || error, null, '  ');
                    ApiErrorHandler.handle(response);
                    return;
                }
                $scope.responseStatus = "success";
                $scope.response = JSON.stringify(data, null, '  ');
                $scope.$evalAsync();
            };
            updateMethod.call(UpdateV2, indexType, $routeParams.core, postData, updateOptions, v2Callback);
        }
    });
